set-base-url.py

```diff
@@ -0,0 +1,12 @@
+import fileinput
+
+
+def main():
+    file_path = "data/base.yaml"
+    for line in fileinput.input(file_path, inplace=True):
+        replace_url = line.replace("docs.ansible.com/", "stage-docsite.readthedocs.io")
+        print(replace_url.strip("\n"))
```

Comment:
Shouldn't this be written back somewhere?

Reply:
From the `help()` of `inputfile` (HTML version: https://docs.python.org/3/library/fileinput.html):

> Optional in-place filtering: if the keyword argument inplace=True is passed to [fileinput.input()](https://docs.python.org/3/library/fileinput.html#fileinput.input) or to the [FileInput](https://docs.python.org/3/library/fileinput.html#fileinput.FileInput) constructor, the file is moved to a backup file and standard output is directed to the input file (if a file of the same name as the backup file already exists, it will be replaced silently).

(TBH `inputfile` is something I've never seen before.)

Reply:
Thanks for the comments!

@webknjaz I was looking at the same docs as Felix pointed out and took it that an explicit backup wouldn't be required. There is a backup parameter that allows you to do something like this:

```
for line in fileinput.input(file_path, inplace=True, backup='.original'):
```

This creates a copy of the file with the original content. I get that's a good practice in certain cases. But in this case we're running this only when deploying to readthedocs. Maybe I don't fully understand your comment but I'm not sure I see the need for the backup here. Cheers.

Reply:
@felixfontein oh, I think I saw that doc once in my life. Though it seems way too magical to me.

@oraNod so my comment was coming from a place of confusion since I didn't realize what the behavior of `fileinput` is, having never used it myself. From the maintainability perspective, this can be hard to follow for a new reader not knowing the context so I'd avoid using it this way. Or maybe I'd wrap into a function communicating that it's basically a pipe into itself better.

One way it would be easier to consume is something like `pathlib.Path(file_path).write_text(pathlib.Path(file_path).read_text(encoding="utf-8").replace("docs.ansible.com/", "stage-docsite.readthedocs.io/"), encoding="utf-8")` that reads the file into memory converts the text and writes the file back to disk. It's small enough not to be concerned about having to process it in chunks.

Though, looking at the PR again, I feel like this shouldn't even be implemented through a separate script.

Instead, it's better to inject the conversion somewhere on the level of `load_page_data()`. The easiest way is to have an argument there for overriding the base domain and making the conversion.

```python
_sentinel_default = object()

def load_page_data(*, base_url=_sentinel_default):
    context_objects = Path("./data").glob("*.yaml")
    context_mapping = {
        file_path.stem: load(file_path.read_text(), Loader=Loader)
        for file_path in context_objects
    }
    if base_url is not _sentinel_default:
        context_mapping['base']['base_url'] = base_url
    return context_mapping
```

A more complicated way could be doing this on the yaml loader level or even defining a custom constructor/type in YAML.

The above is easily usable with partial — it can be passed to the static jinja site generator as follows:
```
load_page_data_callable = partial(load_page_data, base_url=custom_domain_cli_arg) if custom_domain_cli_arg else load_page_data
site.contexts = [(".*.html", load_page_data_callable)]
```

You can then implement passing an argument into nox so that the RTD invocation would be `nox -s build -- "${READTHEDOCS_OUTPUT}"/html https://docsite.readthedocs.io/` with the added benefit of being explicit what the domain is in the configuration place where this information is important.

Reply:
> the RTD invocation would be `nox -s build -- "${READTHEDOCS_OUTPUT}"/html https://docsite.readthedocs.io/`

Better yet — you can reuse the actual URL that is present in the env, making sure that it always corresponds to the RTD project and the PR builds would have the correct value as well:
```console
$ nox -s build -- "${READTHEDOCS_OUTPUT}"/html "${READTHEDOCS_CANONICAL_URL}"
```

Also, if you end up creating two RTD projects connected to this repo, each would have its own build with a proper URL instead of a hardcoded value somewhere in a disconnected script.

Reply:
Thanks for the great suggestion @webknjaz (and for the help figuring out that staticjinja issue). I've updated the PR with the changes. PTAL.

Reply:
Looks great! Marking this thread as resolved.